librad/src/keys.rs

```diff
@@ -15,5 +15,274 @@
+        let sk = ed25519::SecretKey::from_slice(bytes.unsecure())
+            .ok_or(IntoSecretKeyError::InvalidSliceLength)?;
+        Ok(Self::from_secret(sk))
+    }
+
+    fn metadata(&self) -> Self::Metadata {}
+}
+
+// PublicKey
+
+impl PublicKey {
+    pub fn verify(&self, sig: &Signature, data: &[u8]) -> bool {
+        ed25519::verify_detached(sig, &data, self)
+    }
+
+    pub fn from_slice(bs: &[u8]) -> Option<PublicKey> {
+        ed25519::PublicKey::from_slice(&bs).map(PublicKey)
+    }
+
+    pub fn from_bs58(s: &str) -> Option<Self> {
+        let bytes = match bs58::decode(s.as_bytes())
+            .with_alphabet(bs58::alphabet::BITCOIN)
+            .into_vec()
+        {
+            Ok(v) => v,
+            Err(_) => return None,
+        };
+        ed25519::PublicKey::from_slice(&bytes).map(PublicKey)
+    }
+
+    pub fn to_bs58(&self) -> String {
+        bs58::encode(self.0)
+            .with_alphabet(bs58::alphabet::BITCOIN)
+            .into_string()
+    }
+}
+
+impl From<ed25519::PublicKey> for PublicKey {
+    fn from(pk: ed25519::PublicKey) -> Self {
+        Self(pk)
+    }
+}
+
+impl From<SecretKey> for PublicKey {
+    fn from(k: SecretKey) -> Self {
+        k.public()
+    }
+}
+
+impl fmt::Display for PublicKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(&hex::encode(self.as_ref()))
+    }
+}
+
+impl AsRef<[u8]> for PublicKey {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+impl Deref for PublicKey {
+    type Target = ed25519::PublicKey;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+// Signature
+
+pub mod signature {
+    use thiserror::Error;
+
+    #[derive(Debug, Error)]
+    pub enum DecodeError {
+        #[error("Could not decode hex string")]
+        InvalidHex(#[from] hex::FromHexError),
+
+        #[error("Expected signature to be 64 bytes, got: {actual}")]
+        InvalidLength { actual: usize },
+    }
+}
+
+impl Signature {
+    pub fn verify(&self, data: &[u8], pk: &PublicKey) -> bool {
+        ed25519::verify_detached(self, &data, pk)
+    }
+
+    pub fn from_hex_string(s: &str) -> Result<Self, signature::DecodeError> {
+        let bytes = hex::decode(s)?;
+        let buffer = if bytes.len() == 64 {
+            let mut buffer = [0u8; 64];
+            for (i, v) in bytes.iter().enumerate() {
+                buffer[i] = *v;
+            }
+            buffer
+        } else {
+            return Err(signature::DecodeError::InvalidLength {
+                actual: bytes.len(),
+            });
+        };
+        Ok(Self(ed25519::Signature(buffer)))
+    }
+
+    pub fn from_bs58(s: &str) -> Option<Self> {
+        let bytes = match bs58::decode(s.as_bytes())
+            .with_alphabet(bs58::alphabet::BITCOIN)
+            .into_vec()
+        {
+            Ok(v) => v,
+            Err(_) => return None,
+        };
```

Comment:
The indentation threw me off here. I think you could do:
```
let bytes = bs58::decode(s.as_bytes())
     .with_alphabet(bs58::alphabet::BITCOIN)
     .into_vec()
     .ok()?;
```

Reply:
I did not write this. But it should also be obsolete soon (#134 + #121 + #115)